src/bp/nlu-core/training-pipeline.ts

```diff
@@ -181,84 +172,36 @@ const buildVectorsVocab = (intents: Intent<Utterance>[]): _.Dictionary<number[]>
     .value()
 }
 
-export const BuildExactMatchIndex = (input: TrainStep): ExactMatchIndex => {
-  return _.chain(input.intents)
-    .filter(i => i.name !== NONE_INTENT)
-    .flatMap(i =>
-      i.utterances.map(u => ({
-        utterance: u.toString(EXACT_MATCH_STR_OPTIONS),
-        contexts: i.contexts,
-        intent: i.name
-      }))
-    )
-    .filter(({ utterance }) => !!utterance)
-    .reduce((index, { utterance, contexts, intent }) => {
-      index[utterance] = { intent, contexts }
-      return index
-    }, {} as ExactMatchIndex)
-    .value()
-}
-
-const getCustomEntitiesNames = (input: TrainStep): string[] => {
-  return [...input.list_entities.map(e => e.entityName), ...input.pattern_entities.map(e => e.name)]
-}
-
-const TrainIntentClassifier = async (
+const TrainIntentClassifiers = async (
   input: TrainStep,
   tools: Tools,
   progress: progressCB
 ): Promise<_.Dictionary<string>> => {
   debugTraining('Training intent classifier')
-  const customEntities = getCustomEntitiesNames(input)
-  const svmPerCtx: _.Dictionary<string> = {}
 
-  const noneUtts = _.chain(input.intents)
-    .filter(i => i.name === NONE_INTENT) // in case use defines a none intent we want to combine utterances
-    .flatMap(i => i.utterances)
-    .filter(u => u.tokens.filter(t => t.isWord).length >= 3)
-    .value()
+  const { list_entities, pattern_entities, intents, ctxToTrain, nluSeed } = input
 
-  const lo = tools.seededLodashProvider.getSeededLodash()
-  for (let i = 0; i < input.ctxToTrain.length; i++) {
-    const ctx = input.ctxToTrain[i]
-    const trainableIntents = input.intents.filter(
-      i => i.name !== NONE_INTENT && i.contexts.includes(ctx) && i.utterances.length >= MIN_NB_UTTERANCES
-    )
-
-    const nAvgUtts = Math.ceil(_.meanBy(trainableIntents, 'utterances.length'))
-
-    const points = _.chain(trainableIntents)
-      .thru(ints => [
-        ...ints,
-        {
-          name: NONE_INTENT,
-          utterances: lo
-            .chain(noneUtts)
-            .shuffle()
-            .take(nAvgUtts * 2.5) // undescriptible magic n, no sens to extract constant
-            .value()
-        }
-      ])
-      .flatMap(i =>
-        i.utterances.map(utt => ({
-          label: i.name,
-          coordinates: getIntentFeatures(utt, customEntities)
-        }))
-      )
-      .filter(x => !x.coordinates.some(isNaN))
-      .value()
+  const svmPerCtx: _.Dictionary<string> = {}
 
-    if (points.length <= 0) {
-      progress(1 / input.ctxToTrain.length)
-      continue
-    }
-    const svm = new tools.mlToolkit.SVM.Trainer()
+  for (let i = 0; i < ctxToTrain.length; i++) {
```

Comment:
this is pretty much serial, do we gain any kind of speed making this concurrent ?

Reply:
We could, I just didnt want to change the current behavior. As of right now, the intent clf of each ctx is trained in serial.

I will check what I can do in next PRs.